File: common/persons/src/main/java/ru/tensor/sbis/persons/util/PersonNameTemplate.java
package ru.tensor.sbis.persons.util;

import androidx.annotation.Nullable;

import android.text.TextUtils;

/**
 * Created by aa.mironychev on 12.08.16.
 *
 * @deprecated используйте {@link ru.tensor.sbis.design.profile_decl.util.PersonNameTemplate}
 */

@SuppressWarnings("DeprecatedIsStillUsed")
public enum PersonNameTemplate implements NameFormat {
    SURNAME_NAME {
        @Nullable
        @Override
        public String format(String surname, String name, String patronymic) {
            boolean emptyName = TextUtils.isEmpty(name);
            boolean emptySurname = TextUtils.isEmpty(surname);

            if (emptyName) {
                if (emptySurname) {
                    return null;
                }

                return surname;
            }

            if (emptySurname) {
                return name;
            }

            return String.format("%s %s", surname, name);
        }
    },
    SURNAME_N {
        @Nullable
        @Override
        public String format(String surname, String name, String patronymic) {
            boolean emptyName = TextUtils.isEmpty(name);
            boolean emptySurname = TextUtils.isEmpty(surname);

            if (emptyName) {
                if (emptySurname) {
                    return null;
                }

                return surname;
            }

            if (emptySurname) {
                return name;
            }

            return String.format("%s %.1s.", surname, name);
        }
    },
    SURNAME_NAME_PATRONYMIC {
        @Nullable
        @Override
        public String format(String surname, String name, String patronymic) {
            boolean emptyName = TextUtils.isEmpty(name);
            boolean emptySurname = TextUtils.isEmpty(surname);
            boolean emptyPatronymic = TextUtils.isEmpty(patronymic);

            if (emptyName) {
                if (emptySurname) {
                    return null;
                }

                return surname;
            }

            if (emptySurname) {
                return emptyPatronymic
                        ? name
                        : String.format("%s %s", name, patronymic);
            }

            return emptyPatronymic
                    ? String.format("%s %s", surname, name)
                    : String.format("%s %s %s", surname, name, patronymic);
        }
    },
    SURNAME_N_P {
        @Nullable
        @Override
        public String format(String surname, String name, String patronymic) {
            boolean emptyName = TextUtils.isEmpty(name);
            boolean emptySurname = TextUtils.isEmpty(surname);
            boolean emptyPatronymic = TextUtils.isEmpty(patronymic);

            if (emptyName) {
                if (emptySurname) {
                    return null;
                }

                return surname;
            }

            if (emptySurname) {
                return emptyPatronymic
                        ? name
                        : String.format("%s %.1s.", name, patronymic);
            }

            return emptyPatronymic
                    ? String.format("%s %.1s.", surname, name)
                    : String.format("%s %.1s. %.1s.", surname, name, patronymic);
        }
    }
}

File: design/design_profile_decl/src/main/java/ru/tensor/sbis/design/profile_decl/util/PersonNameTemplate.kt
package ru.tensor.sbis.design.profile_decl.util

/**@SelfDocumented */
enum class PersonNameTemplate {
    /**Пример: Иванов И.*/
    SURNAME_N,

    /**Пример: Иванов И. С.*/
    SURNAME_N_P,

    /**Пример: Иванов Иван*/
    SURNAME_NAME,

    /**Пример: Иванов Иван Сергеевич*/
    SURNAME_NAME_PATRONYMIC;

    private class Builder {
        private val list = mutableListOf<String>()

        fun initial(str: String?) = apply {
            if (!str.isNullOrBlank()) list.add(str.take(1).plus("."))
        }

        fun word(str: String?) = apply {
            if (!str.isNullOrBlank()) list.add(str)
        }

        fun build() = list.joinToString(separator = " ")
    }

    /**@SelfDocumented */
    fun format(surname: String?, name: String?, patronymic: String? = "") = Builder().run {
        when (this@PersonNameTemplate) {
            SURNAME_N -> word(surname).initial(name).build()
            SURNAME_N_P -> word(surname).initial(name).initial(patronymic).build()
            SURNAME_NAME -> word(surname).word(name).build()
            SURNAME_NAME_PATRONYMIC -> word(surname).word(name).word(patronymic).build()
        }
    }
}